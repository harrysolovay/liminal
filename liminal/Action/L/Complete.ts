import type { L } from "./L.ts"

export interface Complete<I = any, O = any> {
  kind: "Complete"
  type: L<I, O>
}
