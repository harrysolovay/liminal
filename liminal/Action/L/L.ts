import type { StandardSchemaV1 } from "@standard-schema/spec"
import type { Falsy } from "../../util/Falsy.js"
import type { LMembers } from "./LMembers.js"

export { struct as L } from "./intrinsics/struct.js"

export interface L<I = any, O = I> extends LMembers<I, O>, StandardSchemaV1<I, O> {
  (template: TemplateStringsArray, ...substitutions: Array<string>): this
  (...values: Array<Falsy | string>): this
}

// TODO: use these to minimize signature in cases where I/O is the same
export type I = typeof I
export declare const I: unique symbol
