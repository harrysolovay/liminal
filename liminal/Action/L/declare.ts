import type { Falsy } from "../../util/Falsy.js"
import { isTemplateStringsArray } from "../../util/isTemplateStringsArray.js"
import type { LMembers } from "./LMembers.js"
import { LTag } from "./isL.js"
import type { L } from "./L.js"
import { toJSONSchema } from "./toJSONSchema.js"

export function declare<I, O>(
  self: () => L<I, O> | ((...args: any) => L<I, O>),
  args?: Array<unknown>,
  descriptions: Array<string> = [],
): L<I, O> {
  return Object.assign(annotate, {
    [LTag]: true,
    self,
    args,
    descriptions,
    toJSONSchema,
    *[Symbol.iterator]() {
      return yield {
        kind: "Complete",
        type: this as L<I, O>,
      }
    },
    "~standard": {
      version: 1,
      vendor: "liminal",
      // TODO
      validate: (value) => ({ value: value as O }),
    },
  } satisfies LMembers<I, O>) as never

  function annotate(template: TemplateStringsArray, ...substitutions: Array<string>): L<I, O>
  function annotate(...values: Array<Falsy | string>): L<I, O>
  function annotate(e0: TemplateStringsArray | Falsy | string, ...rest: Array<Falsy | string>): L<I, O> {
    return declare(self, args, [
      ...descriptions,
      ...(isTemplateStringsArray(e0) ? [String.raw(e0, ...rest)] : [e0, ...rest].filter((v): v is string => !!v)),
    ])
  }
}
