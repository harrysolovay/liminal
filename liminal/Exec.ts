import type { Action } from "./Action/Action.js"
import type { Tool } from "./Action/Tool.js"
import { StateReducers } from "./StateReducers/StateReducers.js"
import type { Adapter } from "./Adapter.js"
import type { ActorLike } from "./common/ActorLike.js"
import type { ExecConfig } from "./ExecConfig.js"
import type { ExecState } from "./ExecState.js"
import { unwrapDeferred } from "./util/unwrapDeferred.js"
import type { ExecSpec } from "./ExecSpec.js"

export function Exec<S extends ExecSpec>(adapter: Adapter<S>): Exec<S> {
  return {
    run: async (agent, config) => {
      const reducers = StateReducers(adapter.providerReducers)
      const state: ExecState<S> = {
        config,
        source: agent,
        actor: unwrapDeferred(agent),
        languageModelKey: "default",
        embeddingModelKey: config.models.embedding?.default,
        system: undefined,
        next: undefined,
        parent: undefined,
        handler: config.handler as never,
        messages: [],
        tools: new Set<Tool>(),
      }
      return (await reducers.reduceState(state)) as never
    },
  }
}

export interface Exec<S extends ExecSpec> {
  run: <Y extends Action, R>(agent: ActorLike<Y, R>, config: ExecConfig<S>) => Promise<unknown>
}
