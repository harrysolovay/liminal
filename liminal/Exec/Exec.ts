import type { ActionLike } from "../Action/ActionLike.js"
import type { ActorLike } from "../Actor/ActorLike.js"
import { reduceActor } from "../Actor/reduceActor.js"
import type { ExtractModelAdapters } from "../Config.js"
import { Events } from "../Events.js"
import { Scope } from "../Scope/Scope.js"
import type { ExtractSpec, Spec } from "../Spec.js"
import { unwrapDeferred } from "../util/unwrapDeferred.js"
import type { ExecEnteredEvent, ExecExitedEvent } from "./ExecEvent.js"

export interface Exec<S extends Spec, T> {
  exec: (handler?: (event: ExecEnteredEvent | S["Event"] | ExecExitedEvent<T>) => any) => Promise<Scope<T>>
}

export interface ExecBuilder<S extends Spec, R> {
  models: (models: ExtractModelAdapters<S>) => Exec<S, R>
}

export function Exec<Y extends ActionLike, T, S extends ExtractSpec<Y>>(
  actorLike: ActorLike<Y, T>,
): ExecBuilder<S, T> {
  return {
    models: (models) => ({
      exec: async (handler) => {
        let scope = new Scope(
          models,
          undefined,
          unwrapDeferred(actorLike),
          new Events((inner) => inner, handler),
        )
        scope.events.emit({
          type: "exec_entered",
        })
        scope = await reduceActor(scope)
        scope.events.emit({
          type: "exec_exited",
          result: scope.result,
        })
        return scope
      },
    }),
  }
}
