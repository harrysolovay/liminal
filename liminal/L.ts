import * as AiError from "@effect/ai/AiError"
import * as AiInput from "@effect/ai/AiInput"
import * as AiLanguageModel from "@effect/ai/AiLanguageModel"
import * as AiToolkit from "@effect/ai/AiToolkit"
import * as Effect from "effect/Effect"
import * as Schema from "effect/Schema"
import * as Strand from "./Strand.ts"
import { isTemplateStringsArray } from "./util/isTemplateStringsArray.ts"

export const user: {
  (template: TemplateStringsArray, ...substitutions: Array<unknown>): Effect.Effect<void, never, Strand.Strand>
  (message: string): Effect.Effect<void, never, Strand.Strand>
  (a0: TemplateStringsArray | string, ...aRest: Array<unknown>): Effect.Effect<void, never, Strand.Strand>
} = Effect.fnUntraced(function*(a0, ...aRest) {
  const { messages, onMessage } = yield* Strand.Strand
  const text = isTemplateStringsArray(a0) ? String.raw({ raw: a0 }, ...aRest) : a0
  const message = new AiInput.UserMessage({
    parts: [new AiInput.TextPart({ text })],
  })
  yield* onMessage(message)
  messages.push(message)
})

export const messages: Effect.Effect<Array<AiInput.Message>, never, Strand.Strand> = Effect.map(
  Strand.Strand,
  ({ messages }) => messages,
)

export const setMessages: (
  messages: Iterable<AiInput.Message>,
) => Effect.Effect<Array<AiInput.Message>, never, Strand.Strand> = Effect.fnUntraced(
  function*(messages: Iterable<AiInput.Message>) {
    const strand = yield* Strand.Strand
    const { messages: prev } = strand
    strand.messages = [...messages]
    return prev
  },
)

export const assistant: Effect.Effect<
  string,
  AiError.AiError,
  AiLanguageModel.AiLanguageModel | Strand.Strand
> = Effect.gen(function*() {
  const model = yield* AiLanguageModel.AiLanguageModel
  const { system, messages, tools } = yield* Strand.Strand
  const response = yield* model.generateText({
    system,
    prompt: messages,
    toolkit: AiToolkit.make(...tools ?? []) as never,
  })
  const { text } = response
  yield* append(text)
  return text
})

export const assistantStruct: <O, I extends Record<string, unknown>>(
  schema: Schema.Schema<O, I, never>,
) => Effect.Effect<O, AiError.AiError, AiLanguageModel.AiLanguageModel | Strand.Strand> = Effect.fnUntraced(
  function*(schema) {
    const model = yield* AiLanguageModel.AiLanguageModel
    const { system, messages } = yield* Strand.Strand
    const response = yield* model.generateObject({
      system,
      schema,
      prompt: messages,
    })
    const { value, text } = response
    yield* append(text)
    return value
  },
)

const append = Effect.fnUntraced(function*(text: string) {
  const { messages, onMessage } = yield* Strand.Strand
  const message = new AiInput.AssistantMessage({
    parts: [new AiInput.TextPart({ text })],
  })
  yield* onMessage(message)
  messages.push(message)
})
