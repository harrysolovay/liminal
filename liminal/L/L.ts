export * from "liminal-schema/factories"
export * from "./_infer.ts"
export * from "./_message.ts"
export * from "./assistant.ts"
export * from "./branch.ts"
export * from "./emit.ts"
export * from "./fork.ts"
export * from "./model.ts"
export * from "./system.ts"
export * from "./user.ts"
