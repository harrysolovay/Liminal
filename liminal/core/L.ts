import type { Assertion, DescriptionArg, DescriptionParam, Metadata } from "./Annotation.ts"

export * from "./intrinsics/mod.ts"
export * from "./utility/mod.ts"

export function _<K extends symbol, T = string>(
  key: K,
  serializer?: (value: T) => string,
): DescriptionParam<K, T> {
  return Object.assign(
    (value: T): DescriptionArg<K, T> => ({
      type: "DescriptionArg",
      key,
      value,
      serializer,
    }),
    {
      type: "DescriptionParam" as const,
      key,
    },
  )
}

export function assertion<T, A extends unknown[]>(
  description: string | ((...args: A) => string),
  f?: (value: T, ...args: A) => void | Promise<void>,
): (...args: A) => Assertion<T, A> {
  return (...args) => ({
    type: "Assertion",
    description,
    f,
    args,
  })
}

export function metadata(key: symbol, value: unknown): Metadata {
  return {
    type: "Metadata",
    key,
    value,
  }
}
