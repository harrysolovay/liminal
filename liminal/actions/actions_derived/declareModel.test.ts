import { describe, expect, it } from "bun:test"
import type { LEvent } from "../../events/LEvent.ts"
import { exec } from "../../Exec.ts"
import * as L from "../../L.ts"
import { TestEmbeddingModel } from "../../testing/TestEmbeddingModel.ts"
import { TestLanguageModel } from "../../testing/TestLanguageModel.ts"

describe("Model", () => {
  it("generates the expected event sequence", async () => {
    const events: Array<LEvent> = []
    await exec(function*() {
      const a = yield* L.declareModel("secondary", "language")
      yield* L.branch("fork-key", function*() {
        const b = yield* L.declareModel("child_a", "language")
        const c = yield* L.declareModel("child_b", "embedding")
        yield* b
        yield* c
      })
      const c = yield* L.declareModel("tertiary", "embedding")
      yield* a
      yield* c
    }, {
      default: TestLanguageModel(),
      args: {
        secondary: TestLanguageModel(),
        child_a: TestLanguageModel(),
        child_b: TestEmbeddingModel(),
        tertiary: TestEmbeddingModel(),
      },
      handler(event) {
        events.push(event)
      },
    })
    expect(JSON.stringify(events, null, 2)).toMatchSnapshot()
  })
})
