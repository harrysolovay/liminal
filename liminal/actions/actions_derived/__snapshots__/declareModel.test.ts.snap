// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Model generates the expected event sequence 1`] = `
"[
  {
    "scope": [],
    "index": 0,
    "type": "model_pushed",
    "modelKey": "secondary",
    "modelType": "language"
  },
  {
    "scope": [
      "fork-key"
    ],
    "index": 0,
    "type": "forked"
  },
  {
    "scope": [
      "fork-key"
    ],
    "index": 0,
    "type": "model_pushed",
    "modelKey": "child_a",
    "modelType": "language"
  },
  {
    "scope": [
      "fork-key"
    ],
    "index": 0,
    "type": "model_pushed",
    "modelKey": "child_b",
    "modelType": "embedding"
  },
  {
    "scope": [
      "fork-key"
    ],
    "index": 0,
    "type": "model_removed",
    "modelKey": "child_a",
    "modelType": "language"
  },
  {
    "scope": [
      "fork-key"
    ],
    "index": 0,
    "type": "model_removed",
    "modelKey": "child_b",
    "modelType": "embedding"
  },
  {
    "scope": [
      "fork-key"
    ],
    "index": 0,
    "type": "returned"
  },
  {
    "scope": [],
    "index": 0,
    "type": "model_pushed",
    "modelKey": "tertiary",
    "modelType": "embedding"
  },
  {
    "scope": [],
    "index": 0,
    "type": "model_removed",
    "modelKey": "secondary",
    "modelType": "language"
  },
  {
    "scope": [],
    "index": 0,
    "type": "model_removed",
    "modelKey": "tertiary",
    "modelType": "embedding"
  },
  {
    "scope": [],
    "index": 0,
    "type": "returned"
  }
]"
`;
